private/pkg/manifest/storage.go

```diff
@@ -0,0 +1,158 @@
+func NewFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := New()
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	if walkErr := bucket.Walk(ctx, "", func(info storage.ObjectInfo) (retErr error) {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
+		if err != nil {
+			return err
+		}
+		defer func() { retErr = multierr.Append(retErr, obj.Close()) }()
+		digest, err := digester.Digest(obj)
+		if err != nil {
+			return err
+		}
+		if err := m.AddEntry(path, *digest); err != nil {
+			return err
+		}
+		return nil
+	}); walkErr != nil {
+		return nil, walkErr
+	}
+	return m, nil
+}
+
+// NewBucket takes a manifest and an array of blobs and builds a readable storage bucket that
+// contains the files in the manifest.
+func NewBucket(m Manifest, blobs []Blob) (storage.ReadBucket, error) {
+	digestToBlobs := make(map[string]Blob, len(blobs))
+	for _, b := range blobs {
+		digest := b.Digest().String()
+		_, alreadyPresent := digestToBlobs[digest]
+		if alreadyPresent {
+			return nil, fmt.Errorf("blob with digest %q duplicated", digest)
+		}
+		_, presentInManifest := m.PathsFor(digest)
+		if !presentInManifest {
+			return nil, fmt.Errorf("blob with digest %q is not present in manifest", digest)
+		}
+		digestToBlobs[digest] = b
+	}
+	pathToBlob := make(map[string]Blob, len(blobs))
+	for _, path := range m.Paths() {
+		pathDigest, ok := m.DigestFor(path)
+		if !ok {
+			return nil, fmt.Errorf("path %q not present in manifest", path)
+		}
+		blob, ok := digestToBlobs[pathDigest.String()]
+		if !ok {
+			return nil, fmt.Errorf("manifest path %q with digest %q has no associated blob", path, pathDigest.String())
+		}
+		pathToBlob[path] = blob
+	}
+	return &manifestBucket{
+		manifest:   m,
+		pathToBlob: pathToBlob,
+	}, nil
+}
+
+func (m *manifestBucket) Get(ctx context.Context, path string) (storage.ReadObjectCloser, error) {
+	blob, ok := m.pathToBlob[path]
+	if !ok {
+		return nil, storage.NewErrNotExist(path)
+	}
+	file, err := blob.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &manifestBucketObject{
+		path: path,
+		file: file,
+	}, nil
+}
+
+func (m *manifestBucket) Stat(ctx context.Context, path string) (storage.ObjectInfo, error) {
+	_, ok := m.pathToBlob[path]
+	if !ok {
+		return nil, storage.NewErrNotExist(path)
+	}
+	// storage.ObjectInfo only requires path
+	return &manifestBucketObject{path: path}, nil
+}
+func (m *manifestBucket) Walk(ctx context.Context, prefix string, f func(storage.ObjectInfo) error) error {
+	prefix, err := storageutil.ValidatePrefix(prefix)
+	if err != nil {
+		return err
+	}
+	paths := m.manifest.Paths()
+	sort.Slice(paths, func(i, j int) bool {
+		return paths[i] < paths[j]
+	})
+	for _, path := range paths {
+		if !normalpath.EqualsOrContainsPath(prefix, path, normalpath.Relative) {
+			continue
+		}
+		_, ok := m.pathToBlob[path]
+		if !ok {
+			return storage.NewErrNotExist(path)
+		}
```

Comment:
This is impossible. I'd comment about that. :)

Reply:
Yeah, 2 places in which I thought the same haha, I'll add a comment.

Reply:
Although... if we allow [incomplete manifests](https://github.com/bufbuild/buf/pull/1558/files#r1014506057) it could happen.

Reply:
Done, allowed for incomplete manifests.